File: python/day09/part1.py
with open('day_09_input.txt', 'r') as file:
    disk_map = file.read().strip()

# Parse the disk map to create the initial list of blocks
blocks = []
file_id = 0  # File IDs start from 0
is_file = True  # Indicator to switch between file and free space

i = 0
while i < len(disk_map):
    length = int(disk_map[i])
    # Handle multi-digit lengths if the length is zero
    # As per problem description, '0' represents zero-length file or free space
    # So, we can proceed without special handling for multi-digit lengths
    if is_file:
        blocks.extend([file_id] * length)
        file_id += 1
    else:
        blocks.extend(['.'] * length)
    is_file = not is_file
    i += 1

# Simulate the moving process
while True:
    # Find the leftmost free space
    try:
        lfs = blocks.index('.')
    except ValueError:
        # No free spaces left
        break  # Disk is fully compacted

    # Find the rightmost file block
    try:
        rfb = len(blocks) - 1 - blocks[::-1].index(next(b for b in reversed(blocks) if b != '.'))
    except StopIteration:
        # No file blocks found
        break

    if rfb <= lfs:
        # All file blocks are before the leftmost free space; disk is compacted
        break

    # Move one block from rfb to lfs
    blocks[lfs] = blocks[rfb]
    blocks[rfb] = '.'

# Calculate the checksum
checksum = 0
for position, block in enumerate(blocks):
    if block != '.':
        checksum += position * block

print(checksum)

File: rust/src/day09/part1.rs
pub fn solve(input: &str) -> i64 {
    let mut blocks: Vec<i64> = Vec::new();
    let mut file_id = 0;
    let mut is_file = true;

    // Parse input and create initial blocks
    for c in input.trim().chars() {
        let length = c.to_digit(10).unwrap() as i64;

        if is_file {
            blocks.extend(vec![file_id; length as usize]);
            file_id += 1;
        } else {
            blocks.extend(vec![-1; length as usize]); // Use -1 to represent '.'
        }
        is_file = !is_file;
    }

    // Simulate the moving process
    loop {
        // Find leftmost free space
        let lfs = blocks.iter().position(|&x| x == -1);

        // Find rightmost file block
        let rfb = blocks
            .iter()
            .enumerate()
            .rev()
            .find(|&(_, &x)| x != -1)
            .map(|(i, _)| i);

        match (lfs, rfb) {
            (Some(lfs_pos), Some(rfb_pos)) if rfb_pos > lfs_pos => {
                // Move one block from rfb to lfs
                blocks[lfs_pos] = blocks[rfb_pos];
                blocks[rfb_pos] = -1;
            }
            _ => break, // No more moves possible
        }
    }

    // Calculate checksum
    blocks
        .iter()
        .enumerate()
        .filter(|&(_, &block)| block != -1)
        .map(|(pos, &block)| pos as i64 * block)
        .sum()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_solve() {
        let input = "3132"; // Example input
        assert_eq!(solve(input), 14); // Adjust expected value based on your test case
    }
}
